File: matlab/createGjNetwork.m
function createGjNetwork()
% CREATEGJNETWORK creates a coupled TRN network from coupling parameters and,
%   adjusts Rin of cells in the network
%   recalculates initial conditions for network
%   calculates coupling coefficients (CC) between all cells

rng("shuffle")

load('g_L_func.mat', 'g_L_func')

% vars
namesOfNeurons = {'TRN1','TRN2','TRN3','TRN4','TRN5','TRN6','TRN7','TRN8','TRN9'};
numNeurons = length(namesOfNeurons);
Gc_total = 0.005:0.005:0.03;
numGc = length(Gc_total);
%Gc_sig = 0;
numNets = 1;

Rin_tgt = 7.89;
iDC = -0.3:0.05:0.15;

gj{numGc,numNets}  =[];
g_L{numGc,numNets} =[];
Rin{numGc,numNets} =[];
bias{numGc,numNets}=[];


save_dir = [pwd '/TRN_network'];
if ~isfolder(save_dir)
    mkdir(save_dir)

    save([save_dir '/netVars.mat'], 'namesOfNeurons','Gc_total','numNets',...
                                    'gj', 'g_L', 'Rin', 'bias')

    m=1; n=1;
    save([save_dir '/tmpVars.mat'], 'm','n')
else
    load([save_dir '/tmpVars.mat'], 'm','n')

    load([save_dir,'/netVars.mat'], 'gj', 'g_L', 'Rin', 'bias')
end


% Iterate networks
while m <= numGc
    Gc = Gc_total(m);
    while n <= numNets
        %gj_tmp = constructGJ(numNeurons,Gc,Gc_sig);

        % Ring
        gj_tmp = [0 Gc 0 0 0 0 0 0 Gc;
                  Gc 0 Gc 0 0 0 0 0 0;
                  0 Gc 0 Gc 0 0 0 0 0;
                  0 0 Gc 0 Gc 0 0 0 0;
                  0 0 0 Gc 0 Gc 0 0 0;
                  0 0 0 0 Gc 0 Gc 0 0;
                  0 0 0 0 0 Gc 0 Gc 0;
                  0 0 0 0 0 0 Gc 0 Gc;
                  Gc 0 0 0 0 0 0 Gc 0];

        %Estimate Rin adjust
        g_L_tmp = g_L_func(sum(gj_tmp));

        Rin_tmp = zeros(1,numNeurons);
        bias_tmp= zeros(1,numNeurons);

        save([save_dir '/gj_tmp.mat'], 'gj_tmp')
        save([save_dir '/g_L_tmp.mat'],'g_L_tmp')

        % Rin calculation//adjustment
        adjRin();

        gj(m,n)  = {gj_tmp};
        g_L(m,n) = {g_L_tmp};
        Rin(m,n) = {Rin_tmp};
        bias(m,n)= {bias_tmp};

        save([save_dir,'/netVars.mat'], 'gj', 'g_L', 'Rin', 'bias')

        n = n+1;
        save([save_dir '/tmpVars.mat'],'n')
    end


    m = m+1;
    save([save_dir '/tmpVars.mat'],'m')
end


% Recalculate initial conditions
! julia --profile -e 'using Models.CalcVm;main()'

% Calculate CC
! julia --profile -e 'using Models.CalcCC;main()'


% rm *_tmp.mat


% functions
function adjRin()

    %Calculate Rin
    ! julia --profile -e 'using Models.calcRin;main()'
        %;exit()' ??

    load([save_dir, '/vm_tmp.mat'],'vm')

    for i=1:numNeurons
        [fit1] = fit(iDC',vm(i,:)','poly1');
        Rin_tmp(i) = fit1.p1;
        bias_tmp(i) = (-70-(fit1.p2))/fit1.p1;
    end

    Rin_diff = Rin_tmp-Rin_tgt;

    if any(abs(Rin_diff)./Rin_tgt > 0.01)
    %Adjust leak
    gL_adj = 0.0025*(0.6^counter); % ?? change this based of Rin diff?
    for i=1:numNeurons
        if Rin_diff(i)/Rin_tgt <= 0.01

        elseif Rin(i)>Rin_tgt
            g_L_tmp(i)=g_L_tmp(i)+gL_adj;
        else
            g_L_tmp(i)=g_L_tmp(i)-gL_adj;
        end
    end

    save([save_dir '/g_L_tmp.mat'],'g_L_tmp')

    %Recalculate Rin recursively
    adjRin();

    end

end %adjRin


function [gjMat] = constructGJ(num_TRN,mu_gc,sig_gc)
    %CONSTRUCTGJ generate random coupling within TRN network
    %   Each cell makes 1-3 connections with P 0.45 0.5 0.05 respectively
    %   P synapse will favor cells with no GJs (0.2 bias)
    %   Coupling strength values are gaussian following mu,sig

    gjMat=zeros(num_TRN);
    for i=randperm(num_TRN)
        r=rand;
        if r>=0.95
        numGJ=3;
        elseif r>0.45 && r<0.95
        numGJ=2;
        elseif r<=0.45
        numGJ=1;
        end

        numGJ=numGJ-sum(gjMat(:,i));
        if numGJ<0
            numGJ=0;
        end

        r2=rand(1,9);
        for ii=1:num_TRN
            if gjMat(i,ii)==1 || sum(gjMat(:,ii))>=3
            r2(ii)=0;
            end
            if sum(gjMat(:,ii))==0
            r2(ii)=r2(ii)+0.2;
            end
        end
        r2(i)=0;

        [~,I]=maxk(r2,numGJ);
        gjMat(i,I)=1;

        gjMat=gjMat+gjMat';
        gjMat(gjMat>1)=1;
    end

    gjMat = triu(gjMat);

    gc = sig_gc.*randn(size(gjMat))+mu_gc;
    gjMat = gjMat.*gc;

    gjMat=gjMat+gjMat';
end %constructGJ

end %main

File: networks/createGJnetwork.jl

using Models.TRNnetwork
using OrdinaryDiffEq
using CurveFit
using JLD2

# Run scripts
include("run_scripts/calcRin.jl")
include("run_scripts/calcBias.jl")
include("run_scripts/calcUinit.jl")
include("run_scripts/calcCC.jl")

# Functions
gL_func(Gc) = 0.1 .- 0.9773Gc .+ 6.33Gc.^2
gL_adj(dRin)= 0.0001.*(-0.2243 .- 254.5dRin .+ 35.73dRin.^2 .- 2.59dRin.^3)


# main program
function main()
# vars
namesOfNeurons = ["TRN$i" for i in 1:9]
numNeurons = length(namesOfNeurons)
meanGc  = 0.005:0.005:0.03
numGc   = length(meanGc)
all_sigGc = 0.0:0.001:0.005
numNets = 50

Rin_tgt = 6.239
vm_tgt  = -70

for sigGc in all_sigGc

savepath = joinpath(pwd(),"rand_sig_0_00$(Int(sigGc*1000))")
if ~isdir(savepath)
    mkdir(savepath)

    jldsave(joinpath(savepath,"net_vars.jld2");namesOfNeurons,meanGc,sigGc,numNets)

    gj   = fill(zeros(numNeurons,numNeurons),(numGc,numNets))
    g_L  = fill(zeros(numNeurons),(numGc,numNets))
    Rin  = fill(zeros(numNeurons),(numGc,numNets))

    save_object(joinpath(savepath,"gj.jld2"), gj)
    save_object(joinpath(savepath,"g_L.jld2"),g_L)
    save_object(joinpath(savepath,"Rin.jld2"),Rin)

    saveBlk = 1
    save_object(joinpath(savepath,"saveBlk.jld2"),saveBlk)
else
    gj  = load_object(joinpath(savepath,"gj.jld2"))
    g_L = load_object(joinpath(savepath,"g_L.jld2"))
    Rin = load_object(joinpath(savepath,"Rin.jld2"))

    saveBlk = load_object(joinpath(savepath,"saveBlk.jld2"))
end

# Iterate networks
while saveBlk <= numGc
    Gc = meanGc[saveBlk]
    for n=1:numNets
        gj_tmp = constructGJ(numNeurons,Gc,sigGc)

        #Estimate Rin adjust
        g_L_tmp = gL_func(vec(sum(gj_tmp,dims=1)'))

        # Rin calculation//adjustment
        Rin_tmp = calcRin(namesOfNeurons,gj_tmp,g_L_tmp)

        Rin_diff = Rin_tmp.-Rin_tgt
        while any(abs.(Rin_diff)./Rin_tgt .> 0.001)
            #Adjust leak
            g_L_tmp = g_L_tmp.-gL_adj(Rin_diff)

            #Recalculate Rin
            Rin_tmp = calcRin(namesOfNeurons,gj_tmp,g_L_tmp)

            Rin_diff = Rin_tmp.-Rin_tgt
        end

        gj[saveBlk,n]  = gj_tmp
        g_L[saveBlk,n] = g_L_tmp
        Rin[saveBlk,n] = Rin_tmp
    end

    save_object(joinpath(savepath,"gj.jld2"), gj)
    save_object(joinpath(savepath,"g_L.jld2"),g_L)
    save_object(joinpath(savepath,"Rin.jld2"),Rin)

    saveBlk += 1
    save_object(joinpath(savepath,"saveBlk.jld2"),saveBlk)
end

# Calculate bias current
bias = calcBias(namesOfNeurons,gj,g_L,vm_tgt)

save_object(joinpath(savepath,"bias.jld2"),bias)

# Recalculate Rest
uinit = calcUinit(namesOfNeurons,gj,g_L)

save_object(joinpath(savepath,"uinit.jld2"),uinit)

# Calculate CC
cc = calcCC(namesOfNeurons,gj,g_L,uinit)

save_object(joinpath(savepath,"cc.jld2"),cc)
end

return nothing
end #main

main()

